[#63912] Support multiple authentication provider user links

ba1ash opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->
https://community.openproject.org/work_packages/63912
# What are you trying to accomplish?
<!-- Provide a description of the changes. -->
Change how users relate to auth_providers in a way that supports multiple relations. 
## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

There is a migration of data from `users.identity_url` to a dedicated table `user_auth_provider_links`.
`UserAuthProviderLink` is a related model.
Places where `user.identity_url` is used has been replaced appropriately.
`users.identity_url` marked as deprecated without an alternative at the moment.

**An important behavior change:**
AuthProvider must be created before `users.identity_url` is set.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

ba1ash committed:
[#63912] Support multiple authentication provider user links

https://community.openproject.org/work_packages/63912

ba1ash committed:
Merge branch 'dev' into feature/63912-support-multiple-authentication-provider-user-links

NobodysNightmare commented on app/contracts/users/base_contract.rb:
```diff
@@ -88,6 +91,12 @@ def validate_password_writable
+    def validate_identity_url_writable
+      return if identity_url_writable?
+
+      errors.add(:identity_url, :error_readonly) if model.user_auth_provider_links.any?(&:changed?)
```
:yellow_circle: I am not quite sure about the intended semantics, but things I am wondering:

1. Should the `model.user_auth_provider_links.any?` be part of `identity_url_writable?`?
2. Is `:error_readonly` the correct error code for the problem?
3. Is it okay for an admin to overwrite the identity url even if the auth_provider_links have changed? What happens then?

ba1ash replied:
1. I do not think so. 
  `model.user_auth_provider_links.any?(&:changed?)` is way I used to identify an attempt to change identity_url.
   And `identity_url_writable?` is responsible to check if user is allowed to do the change.
2. I suppose yes, because it is returned when the field is not writable by the user && there is a attempt to change it. `password` field behave in the same way...
3. I think it is okay. What can happen is initialization of new or update of existing `user_auth_provider_link` (depending on `slug` presented in `identity_url`) in memory. And later update in the database should happen.

NobodysNightmare commented on app/controllers/concerns/user_invitation.rb:
```diff
@@ -103,7 +103,7 @@ def clear_tokens(user_id)
 
   def reset_login(user_id)
-    User.where(id: user_id).update_all identity_url: nil
+    UserAuthProviderLink.where(user_id:).destroy_all
```
:green_circle:  Do we require `destroy` here? The previous code used `update_all`, which would update many records via SQL, not triggering callbacks. The counterpart for the new code would be `delete_all`, which also performs the deletion in bulk via SQL, but also doesn't trigger callbacks.

That said, this code only seems to be intended to be used with a single id, so I don't think that the performance considerations are too strong here.

ba1ash replied:
`delete_all` should work just fine. We do not have callbacks to be run here. I'll replace the call with `delete_all`.

NobodysNightmare commented on app/models/auth_provider.rb:
```diff
@@ -29,6 +29,8 @@
 class AuthProvider < ApplicationRecord
   belongs_to :creator, class_name: "User"
 
+  has_many :user_auth_provider_links, dependent: :destroy
```
:green_circle: Because that's possible now:

```suggestion
  has_many :user_auth_provider_links, dependent: :destroy
  has_many :users, through: :user_auth_provider_links
```

NobodysNightmare commented on app/models/auth_provider.rb:
```diff
@@ -55,6 +57,10 @@ def callback_url
   end
 
+  def to_s
+    slug
```
:exclamation: I didn't yet scroll far enough, but is this required for magic later on?

I am asking, because an expectable `to_s` representation of an auth provider would also be the `display_name`, so seeing the slug here is somewhat surprising.

ba1ash replied:
True it is not obvious whether it should be `display_name` or `slug`. There is no strong reasoning behind this change, just desire to skip method call on string interpolation. I will remove the change to avoid confusion.

NobodysNightmare commented on app/models/user.rb:
```diff
@@ -96,6 +96,7 @@ class User < Principal
   has_many :emoji_reactions, dependent: :destroy
   has_many :reminders, foreign_key: "creator_id", dependent: :destroy, inverse_of: :creator
   has_many :remote_identities, dependent: :destroy
+  has_many :user_auth_provider_links, dependent: :destroy
```
:green_circle: Also here, because it's possible:

```suggestion
  has_many :user_auth_provider_links, dependent: :destroy
  has_many :authentication_providers, through: :user_auth_provider_links
```

NobodysNightmare commented on app/services/authentication/omniauth_service.rb:
```diff
@@ -134,28 +141,37 @@ def lookup_or_initialize_user
-        user.identity_url = user_attributes[:identity_url]
-        tok.destroy
+      token.user.tap do |user|
+        if identity_url.present?
```
:yellow_circle: When I tried to understand where `identity_url` is coming from here, I noticed that this is going in circles inside the same class:

1. Here we split the `self.identity_url` into `slug` and `external_id`
2. `self.identity_url` is set from `user_attributes[:identity_url]`
3. user_attributes[:identity_url] is set from `identity_url_from_omniauth`
4. `identity_url_from_omniauth` builds the identity url from the globally available `auth_hash` by joining `:provider` and `:uid`

**tl; dr:** This looks like we could simplify the access patterns by avoiding the intermediate step of an `identity_url`. WDYT?

ba1ash replied:
Yes, it makes sense. I removed `self.identity_url`. Further simplification I suggest to do when we remove `identity_url` entity completely.

NobodysNightmare commented on app/services/authentication/omniauth_service.rb:
```diff
@@ -275,15 +284,20 @@ def build_omniauth_hash_to_user_attributes
+    # Indicates whether OmniAuth::Strategies::Delevoper strategy is used.
+    # https://github.com/omniauth/omniauth/blob/0bcfd5b25bf946422cd4d9c40c4f514121ac04d6/lib/omniauth/strategies/developer.rb
+    # if true:
+    #   identity_url should no be set i
```
Incomplete comment?

ba1ash replied:
Yes

NobodysNightmare commented on db/migrate/20250512114003_move_users_identity_url_to_user_auth_provider_links.rb:
```diff
@@ -0,0 +1,72 @@
+      t.references :auth_provider, null: false, foreign_key: { on_delete: :cascade, on_update: :cascade }
+      t.string :external_id, null: false
+      t.timestamps null: false
+      t.index %i[user_id auth_provider_id], unique: true
```
:yellow_circle: As far as I can tell we should add another index over `%i[auth_provider_id external_id]`, probably also unique.

This is because we query links by `auth_provider` and `exernal_id` frequently. WDYT?

ba1ash replied:
![image](https://github.com/user-attachments/assets/4da9c605-37c1-4ebd-bf51-51b470ac2910)
Yes, it makes sense. The index will be used at least in `jwt_oidc` auth strategy and when filtering in SCIM server API(at least nextcloud scim client does it).
The spec says that client should make sure of uniqueness:
![image](https://github.com/user-attachments/assets/5a7ae01f-d3a8-4720-bf85-59a1f95d2570)

I would like to make sure that if `external_id` is unique if present within certain auth_provider. What would you say?
Other thing is that `externaId` is optional. Then if we have a unique constraint it should allow null duplicates.

ba1ash replied:
I just noticed that I marked `external_id` as mandatory. I suppose in our case it does not much sense to not have `external_id` set.

NobodysNightmare replied:
Yeah. I also recognize that the SCIM specification says that the server (we) does not enforce uniqueness (whatever this means, it's neither "SHOULD NOT enforce" nor "MUST NOT enforce"), I read this PR mostly in the context of SSO and linking SSO providers to a user. In that context it wouldn't make sense to have the same external ID twice for a given auth provider, because it would mean that there are two OpenProject users linked to the same auth provider.

Unless we want to support two users using the same provider (I think we don't), we should ensure uniqueness.

NobodysNightmare commented on db/migrate/20250512114003_move_users_identity_url_to_user_auth_provider_links.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    reversible do |direction|
+      direction.up do
```
:red_circle: Do we have a policy on irreversible data migrations? This data migration will cause the entire migration to be irreversible.

I might want to have a second opinion from @machisuji here. In the end I think I can accept if we can only go forward, but we should be sure that everyone is aware of this and that the change is safe.

ba1ash replied:
It seems like it is okay to have one-way data migration. Other thing is that it is preferable to remove columns with two-step approach using `ignored_columns`.

NobodysNightmare replied:
I agree that the two-step approach is safer. However, now I am wondering whether the current migration approach is safe enough. By not removing data from the `identity_url` column anymore, we do not actively break old web workers anymore. That's good :)

But we at least have a race condition left: Once this migration has been executed, any new `identity_url`s being set by old web workers will become stale and not be used by our new code.

The question is how likely it is that the following happens:
* A user logs in via a new SSO identity for the first time (setting an `identity_url`)
* Timeframe: After this migration has started and before the last web workers died

@machisuji Is that unlikely enough, that we call it an acceptable risk (IMO arguable if the timeframe is less than a minute) or can that timeframe be larger, especially on SaaS or for large customers?

ba1ash replied:
@NobodysNightmare good point. I was just about to raise it. If the time frame can be potentially wide enough then it could be preferable to have 500 errors raised due to absence of `users.identity_url` for some time rather than later have some `identity_url` values not migrated.
From the other side affected users would have to login through SSO again. Not too inconvenient for them, right? ;)

machisuji replied:
Most times the migrations, if there are no bigger changes, take less than a minute across all tenants.

So that makes this scenario relatively unlikely on SaaS, much less even on-prem where migrations usually just take a couple of seconds altogether.

I guess we could do the 2-part migration and prevent new registrations at the same time by adding a constraint to the identity_url column like `ADD CONSTRAINT disable_identity_url CHECK (char_length(identity_url) = 0)`?

edit: or perhaps this would have to be implemented as a pre-insert trigger to only block new values while allowing the existing values

ba1ash replied:
@machisuji it looks good to me. I will add this constraint it if there is no objection from @NobodysNightmare side

NobodysNightmare replied:
With all the new info I have right now, I'd also be fine with leaving this as-is.

@ba1ash is right, that this issue for most people is a matter of having their session "hijacked" one more time, so nothing they will practically notice. Also @machisuji's timeline sounds safe and fine to me.

Of course feel free to add additional safe guards, my main concern regarding those is, that they also have to work correctly to not risk breaking anything by merely existing. And since such bugs would only happen around migrations, there is extra difficulty in QAing them. So my preference right now would be to keep it simple.

ba1ash replied:
@machisuji after additional chat with Jan I think it is better to go forward with simple approach(without additional constraint or trigger)

machisuji replied:
@ba1ash Yes I think you both are right. There is quite limited time when invalid records could be created. On SaaS it's even impossible since we would put the tenants into maintenance mode for this kind of migration. So let's go ahead with the simple approach.

NobodysNightmare commented on docs/api/apiv3/components/schemas/user_model.yml:
```diff
@@ -83,10 +83,12 @@ allOf:
           - 'null'
         description: |-
           User's identity_url for OmniAuth authentication.
+          **Deprecated:** It will be removed in the near future. Use `UserAuthProviderLink` API instead.
```
:yellow_circle:  Do I see it correctly, that this API is not added in the scope of this PR?

If so, we should change that description. E.g. not adding the second sentence above.

ba1ash replied:
Yes, there is no API added here. I'll remove the sentence.

NobodysNightmare commented on lib_static/open_project/authentication/strategies/warden/jwt_oidc.rb:
```diff
@@ -30,7 +30,7 @@ def authenticate!
                 end
 
-                user = User.find_by(identity_url: "#{provider.slug}:#{payload['sub']}")
+                user = UserAuthProviderLink.where(auth_provider: provider, external_id: payload["sub"]).first&.user
```
:green_circle: 

```suggestion
                user = UserAuthProviderLink.find_by(auth_provider: provider, external_id: payload["sub"])&.user
```

Or maybe even better:

```suggestion
                user = provider.user_auth_provider_links.find_by(external_id: payload["sub"])&.user
```

NobodysNightmare commented on spec/factories/user_factory.rb:
```diff
@@ -65,7 +65,17 @@
+
+        auth_provider = AuthProvider.find_by(slug:) || create(:oidc_provider, slug:)
+        user.user_auth_provider_links.create!(auth_provider:,
+                                              external_id: external_id)
```
:green_circle: 
```suggestion
                                              external_id:)
```

ba1ash committed:
[#63912] React on PR comments.

Important changes:
1. Use ignored_columns to try to avoid downtime.
2. Add unique constraint for auth_provider_id+external_id.

Co-authored-by: Jan Sandbrink <453584+username@users.noreply.github.com>

ba1ash committed:
[#63912] Try to fix migrations CI run.

NobodysNightmare commented:
FYI: From the discussions we had, this looks fine. I'll give it a manual testing round beginning of next week before approving.

NobodysNightmare reviewed (approved):
Manually trying it out also revealed no further issues... LGTM!

ba1ash committed:
Merge branch 'dev' into feature/63912-support-multiple-authentication-provider-user-links
